v2/test/workload_api.go

```diff
@@ -0,0 +1,321 @@
+}
+
+func (r *X509SVIDResponse) ToProto(tb testing.TB) *workload.X509SVIDResponse {
+	bundle := derBlobFromCerts(r.Bundle)
+
+	pb := &workload.X509SVIDResponse{
+		FederatedBundles: make(map[string][]byte),
+	}
+	for _, svid := range r.SVIDs {
+		// The workload API should always respond with at one certificate and a
+		// private key but making this optional here is needed for some test
+		// flexibility.
+		var spiffeID string
+		if len(svid.CertChain) > 0 && len(svid.CertChain[0].URIs) > 0 {
+			spiffeID = svid.CertChain[0].URIs[0].String()
+		}
+		var keyDER []byte
+		if svid.Key != nil {
+			var err error
+			keyDER, err = x509.MarshalPKCS8PrivateKey(svid.Key)
+			require.NoError(tb, err)
+		}
+		pb.Svids = append(pb.Svids, &workload.X509SVID{
+			SpiffeId:      spiffeID,
+			X509Svid:      derBlobFromCerts(svid.CertChain),
+			X509SvidKey:   keyDER,
+			Bundle:        bundle,
+			FederatesWith: svid.FederatesWith,
+		})
+	}
+	for k, v := range r.FederatedBundles {
+		pb.FederatedBundles[k] = derBlobFromCerts(v)
+	}
+
+	return pb
+}
+
+type bundleDoc struct {
+	jose.JSONWebKeySet
+	Sequence    uint64 `json:"spiffe_sequence,omitempty"`
+	RefreshHint int    `json:"spiffe_refresh_hint,omitempty"`
+}
+
+func (w *WorkloadAPI) fetchX509SVID(_ *workload.X509SVIDRequest, stream workload.SpiffeWorkloadAPI_FetchX509SVIDServer) error {
+	if err := checkHeader(stream.Context()); err != nil {
+		return err
+	}
+	ch := make(chan *workload.X509SVIDResponse, 1)
+	w.mu.Lock()
+	w.x509Chans[ch] = struct{}{}
+	resp := w.respX509
+	w.mu.Unlock()
+
+	defer func() {
+		w.mu.Lock()
+		delete(w.x509Chans, ch)
+		w.mu.Unlock()
+	}()
+
+	sendResp := func(resp *workload.X509SVIDResponse) error {
+		if resp == nil {
+			return noSvidError
+		}
+		return stream.Send(resp)
+	}
+
+	if err := sendResp(resp); err != nil {
+		return err
+	}
+	for {
+		select {
+		case resp := <-ch:
+			if err := sendResp(resp); err != nil {
+				return err
+			}
+		case <-stream.Context().Done():
+			return stream.Context().Err()
+		}
+	}
+}
+
+func (w *WorkloadAPI) fetchJWTSVID(ctx context.Context, req *workload.JWTSVIDRequest) (*workload.JWTSVIDResponse, error) {
+	if err := checkHeader(ctx); err != nil {
+		return nil, err
+	}
+	if len(req.Audience) == 0 {
+		return nil, errors.New("no audience")
+	}
+	//TODO: review if this is the actual error in the implementation
+	if w.respJWT == nil {
+		return nil, noSvidError
+	}
+
+	return w.respJWT, nil
+}
+
+func (w *WorkloadAPI) fetchJWTBundles(_ *workload.JWTBundlesRequest, stream workload.SpiffeWorkloadAPI_FetchJWTBundlesServer) error {
+	if err := checkHeader(stream.Context()); err != nil {
+		return err
+	}
+	ch := make(chan map[string][]byte, 1)
+	w.mu.Lock()
+	w.jwtBundlesChans[ch] = struct{}{}
+	resp := w.respJWTBundles
+	w.mu.Unlock()
+
+	defer func() {
+		w.mu.Lock()
+		delete(w.jwtBundlesChans, ch)
+		w.mu.Unlock()
+	}()
+
+	sendResp := func(resp *workload.JWTBundlesResponse) error {
+		if resp == nil {
```

Comment:
This nil check isn't going to work... the response pointer is always non-nil (it's the bundles that can be empty according to the code).